pkg/covermerger/file_line_merger.go

```diff
@@ -18,33 +22,62 @@ func makeFileLineCoverMerger(
 		return &DeletedFileLineMerger{}
 	}
 	a := &FileLineCoverMerger{
-		rbcToFile: fileVersions,
-		baseFile:  baseFile,
-		hitCounts: make(map[int]int),
-		matchers:  make(map[RepoBranchCommit]*LineToLineMatcher),
+		rbcToFile:  fvs,
+		baseFile:   baseFile,
+		hitCounts:  make(map[int]int),
+		matchers:   make(map[RepoBranchCommit]*LineToLineMatcher),
+		lostFrames: map[RepoBranchCommit]int64{},
 	}
-	for rbc, fileVersion := range fileVersions {
-		a.matchers[rbc] = makeLineToLineMatcher(fileVersion, baseFile)
+	for rbc, fv := range fvs {
+		a.matchers[rbc] = makeLineToLineMatcher(fv.content, baseFile)
 	}
 	return a
 }
 
+// freshestRBC returns RepoBranchCommit with the last modified non-empty file.
+func freshestRBC(fvs fileVersions) *RepoBranchCommit {
+	var res *RepoBranchCommit
+	var resLastUpdated time.Time
+	for rbc, fv := range fvs {
+		if fv.content == "" {
+			continue
+		}
+		if res == nil || resLastUpdated.Before(fv.lastUpdated) {
+			res = &rbc
+			resLastUpdated = fv.lastUpdated
+		}
+	}
+	return res
+}
+
 type FileLineCoverMerger struct {
-	rbcToFile map[RepoBranchCommit]string
-	baseFile  string
-	hitCounts map[int]int
-	matchers  map[RepoBranchCommit]*LineToLineMatcher
+	rbcToFile  fileVersions
+	baseFile   string
+	hitCounts  map[int]int
+	matchers   map[RepoBranchCommit]*LineToLineMatcher
+	lostFrames map[RepoBranchCommit]int64
 }
 
-func (a *FileLineCoverMerger) AddRecord(rbc RepoBranchCommit, arch string, f Frame, hitCount int) {
+func (a *FileLineCoverMerger) AddRecord(rbc RepoBranchCommit, arch string, f *Frame, hitCount int) {
+	if a.matchers[rbc] == nil {
+		if hitCount > 0 {
+			a.lostFrames[rbc]++
+		}
+		return
+	}
 	if targetLine := a.matchers[rbc].SameLinePos(f.StartLine); targetLine != -1 {
```

Comment:
The code assumes some files may be deleted and the tests follow this assumption.
It also assumes we can't receive signals from the empty files.
It may be a problem if you're trying to use old workdir to process fresh signals. In this case signals exist but the commit doesn't exist.
We have signals from forbidden commits and the tool report this as a summary.